src/main/kotlin/app/Analytics.kt

```diff
@@ -0,0 +1,95 @@
+package app
+
+import com.github.kittinunf.fuel.core.FuelManager
+import com.github.kittinunf.fuel.core.Method
+import com.github.kittinunf.fuel.core.Request
+
+typealias Param = Pair<String, String>
+
+/**
+ * Google Analytics events tracking.
+ */
+object Analytics {
+    private val IS_ENABLED = BuildConfig.IS_GA_ENABLED
+    private val BASE_PATH = BuildConfig.GA_BASE_URL
+    private val PROTOCOL_VERSION = "1"
+    private val TRACKING_ID = BuildConfig.GA_TRACKING_ID
+    private val DATA_SOURCE = "app"
+
+    private val HIT_PAGEVIEW = "pageview"
+    private val HIT_EXCEPTION = "exception"
+
+    private val fuelManager = FuelManager()
+
+    var uuid: String = ""  // Should be set on start of the app.
+    var username: String = ""  // Should be set on success authorization.
+
+    init {
+        fuelManager.basePath = BASE_PATH
+    }
+
+    private fun post(params: List<Param>): Request {
+        return fuelManager.request(Method.POST, "/collect", params)
+    }
+
+    private fun trackEvent(event: String, params: List<Param> = listOf()) {
+        if (!IS_ENABLED || (username.isEmpty() && uuid.isEmpty())) {
+            return
+        }
+
+        val idParams = mutableListOf<Param>()
+        if (uuid.isNotEmpty()) {
+            idParams.add("cid" to uuid)
+        }
+        if (username.isNotEmpty()) {
+            idParams.add("uid" to username)
+        }
+
+        val defaultParams = listOf("v" to PROTOCOL_VERSION,
+                                   "tid" to TRACKING_ID,
+                                   "ds" to DATA_SOURCE,
+                                   "t" to HIT_PAGEVIEW,
+                                   "dp" to event)
+
+        try {
+            // Send event to GA with united params.
+            post(params + defaultParams.filter { !params.contains(it) }
+                + idParams).responseString()
+        } catch (e: Throwable) {
+            // Never fail from analytics.
+        }
+    }
+
+    fun trackStart() {
+        trackEvent("/start")
+    }
+
+    fun trackAuth() {
+        trackEvent("/auth")
+    }
+
+    fun trackConfigSetup() {
+        trackEvent("/config/setup")
+    }
+
+    fun trackConfigChanged() {
+        trackEvent("/config/changed")
+    }
+
+    fun trackHashingRepoSuccess() {
+        trackEvent("/hashing/repo/success")
+    }
+
+    fun trackHashingSuccess() {
+        trackEvent("/hashing/success")
+    }
+
+    fun trackError() {
+        // TODO(anatoly): Send message with exd, consider limit 150 bytes.
+        trackEvent("/error", listOf("t" to HIT_EXCEPTION))
```

Comment:
there should be some sort sort of error code as an argument for trackError, e.g.

/virtual/app/error/out-of-disk-space

Reply:
Todo is about it.